src/JuMP.jl

```diff
@@ -53,38 +53,34 @@ include("utils.jl")
 const _MOIVAR = MOI.VariableIndex
 const _MOICON{F,S} = MOI.ConstraintIndex{F,S}
 
-# OptimizerFactory was deprecated in JuMP 0.21 and should be removed when
-# with_optimizer is removed.
-struct OptimizerFactory
-    # The constructor can be
-    # * `Function`: a function, or
-    # * `DataType`: a type, or
-    # * `UnionAll`: a type with missing parameters.
-    constructor
-    args::Tuple
-    kwargs # type changes from Julia v0.6 to v0.7 so we leave it untyped for now
-end
-
-function with_optimizer(constructor,
-                        args...; kwargs...)
+function with_optimizer(constructor; kwargs...)
+    if isempty(kwargs)
+        deprecation_message = """
+with_optimizer is deprecated, replace `with_optimizer(Ipopt.Optimizer)` by `Ipopt.Optimizer`.
+"""
+        Base.depwarn(deprecation_message, :with_optimizer)
+        return constructor
+    else
+        deprecation_message = """
+`with_optimizer` is deprecated, replace `with_optimizer(Ipopt.Optimizer, max_cpu_time=60.0)`
+by `MOI.OptimizerWithAttributes(Ipopt.Optimizer, "max_cpu_time" => 60.0)`.
+"""
+        Base.depwarn(deprecation_message, :with_optimizer_kw)
+        params = [MOI.RawParameter(string(kw.first)) => kw.second for kw in kwargs]
+        return MOI.OptimizerWithAttributes(constructor, params)
+    end
+end
+function with_optimizer(constructor, args...; kwargs...)
     deprecation_message = """
-with_optimizer is deprecated. The examples below demonstrate how to update to the new syntax:
-- 'with_optimizer(Ipopt.Optimizer)' becomes 'Ipopt.Optimizer'.
-- 'set_optimizer(model, with_optimizer(Ipopt.Optimizer, print_level=1, tol=1e-5))' becomes 'set_optimizer(model, Ipopt.Optimizer); set_parameters(model, \"print_level\" => 1, \"tol\" => 1e-5)'.
-- In rare cases where an argument must be passed to the constructor, use an anonymous function. For example, 'env = Gurobi.Env(); set_optimizer(model, with_optimizer(Gurobi.Optimizer, env))' becomes 'env = Gurobi.Env(); set_optimizer(model, () -> Gurobi.Optimizer(env))'.
-    """
-    Base.depwarn(deprecation_message, :with_optimizer)
+`with_optimizer` is deprecated, replace `with_optimizer(Ipopt.Optimizer, args...)` by `() -> Ipopt.Optimizer(args...)`.
```

Comment:
This message implies that keyword arguments should be passed to the constructor, but we're trying to deprecate that as well.

Reply:
This is positional for arguments

Reply:
If you call `with_optimizer(Ipopt.Optimizer, arg1, kwarg1 = arg2)` then you'll see this message that could easily be understood as implying you should write `() -> Ipopt.Optimizer(arg1, kwarg1 = arg2)`.

Reply:
You'll see the other warning too because we redirect to with_optimizer

Reply:
Seeing two deprecation warnings, neither of which provides clear guidance for your situation, isn't a good user experience.